src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -26,10 +29,23 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
     [ExportLspMethod(LSP.Methods.TextDocumentCompletionName)]
     internal class CompletionHandler : AbstractRequestHandler<LSP.CompletionParams, LSP.CompletionItem[]>
     {
+        private readonly ImmutableHashSet<string> _csTriggerCharacters;
+        private readonly ImmutableHashSet<string> _vbTriggerCharacters;
+
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public CompletionHandler(ILspSolutionProvider solutionProvider) : base(solutionProvider)
+        public CompletionHandler(
+            ILspSolutionProvider solutionProvider,
+            [ImportMany] IEnumerable<Lazy<CompletionProvider, CompletionProviderMetadata>> completionProviders)
+            : base(solutionProvider)
         {
+            var csCompletionProviders = completionProviders?.Where(lz => lz.Metadata.Language == LanguageNames.CSharp).ToImmutableArray();
```

Comment:
I don't think you need the ? here

Can probably just do completionProviders.Where(language name equal).SelectMany(GetTriggerChars).Select(tostring).ToImmutableHashSet()

Reply:
Would this break TypeScript at all? Do they use this CompletionHandler at all (via TypeScriptHandlerShim)?

Reply:
there should be completion providers anyway since they're part of the vs mef composition.  But speaking of TS, you may need to update the check to not look for matching trigger chars when in a TS document, otherwise the completion there may break.  Should be able to verify by just livesharing on a web project with a TS file.

Reply:
I see, thanks! I believe the check specifically only checks C#/VB, but I'll double check TS to make sure nothing is broken.

Reply:
I tested with a Nodejs .ts file via Liveshare and completion seems to be working fine, so I'll go ahead and merge now.
Thanks for the review!